flink-ml-lib/src/main/java/org/apache/flink/ml/classification/naivebayes/NaiveBayes.java

```diff
@@ -0,0 +1,377 @@
+        @Override
+        public void flatMap(
+                Tuple2<Vector, Integer> value,
+                Collector<Tuple4<Integer, Integer, Double, Double>> collector) {
+            Preconditions.checkNotNull(value.f1);
+            for (int i = 0; i < value.f0.size(); i++) {
+                collector.collect(new Tuple4<>(value.f1, i, value.f0.get(i), 1.0));
+            }
+        }
+    }
+
+    /**
+     * Function that aggregates entries of feature value and weight into maps.
+     *
+     * <p>Input records should have the same label value and feature column index.
+     *
+     * <p>Input records are tuples with the following fields in order:
+     *
+     * <ul>
+     *   <li>label value
+     *   <li>feature column index
+     *   <li>feature value
+     *   <li>weight
+     * </ul>
+     *
+     * <p>Output records are tuples with the following fields in order:
+     *
+     * <ul>
+     *   <li>label value
+     *   <li>feature column index
+     *   <li>map of (feature value, weight)
+     * </ul>
+     */
+    private static class GenerateFeatureWeightMapFunction
+            implements AggregateFunction<
+                    Tuple4<Integer, Integer, Double, Double>,
+                    Tuple3<Integer, Integer, Map<Double, Double>>,
+                    Tuple4<Integer, Integer, Map<Double, Double>, Double>> {
+
+        @Override
+        public Tuple3<Integer, Integer, Map<Double, Double>> createAccumulator() {
+            return new Tuple3<>(0, -1, new HashMap<>());
+        }
+
+        @Override
+        public Tuple3<Integer, Integer, Map<Double, Double>> add(
+                Tuple4<Integer, Integer, Double, Double> value,
+                Tuple3<Integer, Integer, Map<Double, Double>> acc) {
+            acc.f0 = value.f0;
+            acc.f1 = value.f1;
+            acc.f2.put(value.f2, value.f3);
+            return acc;
+        }
+
+        @Override
+        public Tuple4<Integer, Integer, Map<Double, Double>, Double> getResult(
+                Tuple3<Integer, Integer, Map<Double, Double>> acc) {
+            double weightSum = acc.f2.values().stream().mapToDouble(Double::doubleValue).sum();
+            return new Tuple4<>(acc.f0, acc.f1, acc.f2, weightSum);
+        }
+
+        @Override
+        public Tuple3<Integer, Integer, Map<Double, Double>> merge(
+                Tuple3<Integer, Integer, Map<Double, Double>> acc0,
+                Tuple3<Integer, Integer, Map<Double, Double>> acc1) {
+            Preconditions.checkArgument(acc0.f1 != -1);
+            acc0.f2.putAll(acc1.f2);
+            return acc0;
+        }
+    }
+
+    /**
+     * Function that aggregates maps under the same label into arrays.
+     *
+     * <p>Length of the generated array equals to the number of feature columns.
+     *
+     * <p>Input records are tuples with the following fields in order:
+     *
+     * <ul>
+     *   <li>label value
+     *   <li>feature column index
+     *   <li>map of (feature value, weight)
```

Comment:
It seems that the 4th field of the input is missing?

Reply:
Yes. I'll fix it.